File: dynamicProgramming/MaxSubArray.java
import java.io.*;
import java.util.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

// Solution to https://www.hackerrank.com/challenges/maxsubarray

public class Solution {

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int T = in.nextInt();
        for(int i = 0; i < T; i++){
            int N = in.nextInt();
            int[] arr = new int[N];
            for(int j = 0; j < N; j++){
                arr[j] = in.nextInt();
            }
            int maxEndingHere = arr[0];
            int max = arr[0];
            int posSum = Math.max(arr[0], 0);
            int maxInt = arr[0];
            for(int j = 1; j < N; j++){
                maxEndingHere = Math.max(maxEndingHere + arr[j], 0);
                max = Math.max(max, maxEndingHere);
                if(maxInt < arr[j]) maxInt = arr[j];
                if(arr[j] > 0) posSum+= arr[j];
            }
            int answer = maxInt > 0 ? posSum : maxInt;
            if(posSum == 0) max = maxInt;
            System.out.println(max + " " + answer);
        }
    }
}

File: algorithms/dynamicProgramming/max_sub_array.py
# Solution to https://www.hackerrank.com/challenges/maxsubarray

t = int(input())
for _ in range(t):
    n = int(input())
    a = [int(x) for x in input().split()]

    first = 0
    second = 0
    maxVal = a[0]
    maxEncountered = 0

    for i in a:
        maxEncountered = max(maxEncountered + i, 0)
        first = max(maxEncountered, first)

        if i > 0:
            second = second + i
        maxVal = max(maxVal, i)

    if maxVal < 0:
        second = maxVal
        first = maxVal

    print(str(first) + " " + str(second))
